### DISCUSSION TITLE
How can I build the SN server in my local network?

### DISCUSSION THREAD

**Discussion by streetycat** [buckyos/CYFS#221] [Q&A] [answered]
I have a LAN, I want to build the `CYFS` for it, I have start the [`MetaChain` following this discussion](https://github.com/buckyos/CYFS/discussions/191).

But how can I build the `SN` for it?

**Comment by lurenpluto** [comment]
In cyfs system, SN server is free to choose or even build your own, this is completely dependent on the user, for the case that must use SN (device does not have WAN address), you can configure SN in several ways:

1. use the system default SN server
cyfs service itself has several default SN servers, if the user does not have additional SN configuration, then the default SN will be used
	
2. choose a SN service provider
In the future, there should be various SN service providers, and there may be free or paid depending on the specific situation.
	
3. Build your own SN server
Advanced users can DIY their own SN server, and do the corresponding configuration on OOD

For the last two configurations, it should be noted that if you want to connect to a device, then you must know the other party's SN, so when your own OOD is configured with a custom SN server, then you have to tell others so that they can connect to you (your own OOD device does not have a WAN address)

About how to build your own SN server, I hope @jing-git @photosssa  can provide some tutorials

**Comment by jing-git** [comment]
Hello, you can refer to the following steps to deploy an SN service:

1. Build sn miner server & desc tool
cargo build --release -p sn-miner-rust
cargo build --release -p desc-tool

2. Use desc-tool to reate sn-miner's desc files
For example, sn-miner's IP is A.B.C.D:
./desc-tool create device -c server -d sn_server -e "W4tcpA.B.C.D:8060;W4udpA.B.C.D:8060"
Result:
INFO [desc_tool::desc] succ encode secret to 5aSixgMZAZoU46jdED7urLMrYe7dRzaMNxktFJTTUVPT.sec
INFO [desc_tool::desc] success encode peerdesc to 5aSixgMZAZoU46jdED7urLMrYe7dRzaMNxktFJTTUVPT.desc

3. Rename desc files
mv 5aSixgMZAZoU46jdED7urLMrYe7dRzaMNxktFJTTUVPT.desc sn-miner.desc
mv 5aSixgMZAZoU46jdED7urLMrYe7dRzaMNxktFJTTUVPT.sec sn-miner.sec

4. Put the sn-miner-rust & desc files to the sn-server, and put them in the same directory

5. Start the sn-miner service
For example, the sn-miner-rust & desc files in the /root/snm/:
./sn-miner-rust -d /root/snm/sn-miner

6. Run your CYFS client with sn-miner.desc, enjoy it :)

**Reply by streetycat** [reply, marked as answer]
Thank you, I will try it.

**Comment by streetycat** [comment]
I think, I should put `sn-miner.desc` to the `MetaChain` before it's used by the `CYFS` client. otherwise the `sn` will cannot be found in the client.

```
cyfs-meta-client putdesc -c=sn-miner -d=sn-miner.desc 0 0
```

And, update the `sn` in `device.desc` of the `CYFS` client.

```
desc-tool modify device.desc -s=${sn-device-id};
```

And, put the updated `device.desc` to `MetaChain` for other nodes to find it.

```
cyfs-meta-client putdesc -c=device -d=device.desc 0 0
```
